gst/mqtt/mqttsrc.c

```diff
@@ -0,0 +1,853 @@
+}
+
+/**
+  * @brief A callback to handle the post-processing of the delivered message
+  * @todo Fill the function body
+  */
+static void
+cb_mqtt_on_delivery_complete (void *context, MQTTAsync_token token)
+{
+
+}
+
+/**
+  * @brief A callback to handle the arrived message
+  */
+static int
+cb_mqtt_on_message_arrived (void *context, char *topic_name, int topic_len,
+    MQTTAsync_message * message)
+{
+  const int size = message->payloadlen;
+  guint8 *data = message->payload;
+  GstMQTTMessageHdr *mqtt_msg_hdr;
+  GstMapInfo hdr_map_info;
+  GstMemory *recieved_mem;
+  GstMemory *hdr_mem;
+  GstBuffer *buffer;
+  GstMqttSrc *self;
+  gsize offset;
+  guint i;
+
+  self = GST_MQTT_SRC_CAST (context);
+  recieved_mem = gst_memory_new_wrapped (0, data, size, 0, size, message,
+      (GDestroyNotify) cb_memory_wrapped_destroy);
+  if (!recieved_mem) {
+    self->err = g_error_new (self->gquark_err_tag, ENODATA,
+        "%s: failed to wrap the raw data of recieved message in GstMemory: %s",
+        __func__, g_strerror (ENODATA));
+    return TRUE;
+  }
+
+  mqtt_msg_hdr = _extract_mqtt_msg_hdr_from (recieved_mem, &hdr_mem,
+      &hdr_map_info);
+  if (!mqtt_msg_hdr) {
+    self->err = g_error_new (self->gquark_err_tag, ENODATA,
+        "%s: failed to extract header information from recieved message: %s",
+        __func__, g_strerror (ENODATA));
+    goto ret_unref_recieved_mem;
+  }
+
+  buffer = gst_buffer_new ();
+  offset = GST_MQTT_LEN_MSG_HDR;
+  for (i = 0; i < mqtt_msg_hdr->num_mems; ++i) {
+    GstMemory *each_memory;
+    int each_size;
+
+    each_size = mqtt_msg_hdr->size_mems[i];
+    each_memory = gst_memory_share (recieved_mem, offset, each_size);
+    gst_buffer_append_memory (buffer, each_memory);
+    offset += each_size;
+  }
+
+  g_async_queue_push (self->aqueue, buffer);
+
+  gst_memory_unmap (hdr_mem, &hdr_map_info);
+  gst_memory_unref (hdr_mem);
+
+ret_unref_recieved_mem:
+  gst_memory_unref (recieved_mem);
+
+  return TRUE;
+}
+
+/**
+  * @brief A callback invoked when destroying the GstMemory which wrapped the arrived message
+  */
+static void
+cb_memory_wrapped_destroy (void *p)
+{
+  MQTTAsync_message *msg = p;
+
+  MQTTAsync_freeMessage (&msg);
+}
+
+/**
+  * @brief A callback invoked when the connection is established
+  */
+static void
+cb_mqtt_on_connect (void *context, MQTTAsync_successData * response)
+{
+  GstMqttSrc *self = GST_MQTT_SRC (context);
+  int ret;
+
+  g_mutex_lock (&self->mqtt_src_mutex);
+  self->is_connected = TRUE;
+  g_cond_broadcast (&self->mqtt_src_gcond);
+  g_mutex_unlock (&self->mqtt_src_mutex);
+
+  /** @todo Support QoS option */
+  ret = MQTTAsync_subscribe (self->mqtt_client_handle, self->mqtt_topic, 1,
+      &self->mqtt_respn_opts);
+  if (ret != MQTTASYNC_SUCCESS) {
+    g_printerr ("Failed to start subscribe, return code %d\n", ret);
+    return;
+  }
+
+  g_mutex_lock (&self->mqtt_src_mutex);
+  self->is_subscribed = TRUE;
+  g_mutex_unlock (&self->mqtt_src_mutex);
```

Comment:
Q. No need to broadcast?

Reply:
No, the effective flag changing will be done either ```cb_mqtt_on_subscribe``` or ```cb_mqtt_on_subscribe_failure```.